Assets/Scripts/PlayerMovement.cs

```diff
@@ -15,11 +20,33 @@ private void Start()
     void Update()
     {
         Vector3 move = new Vector3(Input.GetAxis("Horizontal"), 0, Input.GetAxis("Vertical"));
-        controller.Move(move * Time.deltaTime * playerSpeed);
+        controller.Move(move * Time.deltaTime * playerSpeed * GetMovementCoef());
 
         if (move != Vector3.zero)
         {
             gameObject.transform.forward = move;
         }
     }
+
+    private float GetMovementCoef()
+    {
+        int layerMask = 1 << 9; // layer of map tile
+
+        if (Physics.Raycast(transform.position, Vector3.down, out RaycastHit hit, 3, layerMask))
+        {
+            switch (hit.collider.gameObject.tag)
+            {
+                case "Grass":
+                    return grassMovementCoef;
+                case "Bush":
+                    return bushMovementCoef;
+                case "Tree":
+                    return treeMovementCoef;
+                case "Water":
+                    return waterMovementCoef;
+            }
```

Comment:
Add default case, what if by accident we hit no layer?

Reply:
This is not needed. If non of the cases match, it just falls through the switch statement and nothing happens. You can read about it here: [https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch)